src/Sylius/Component/Promotion/Generator/PercentageGenerationPolicy.php

```diff
@@ -68,6 +68,11 @@ private function calculatePossibleGenerationAmount(PromotionCouponGeneratorInstr
             $instruction->getSuffix()
         );
 
-        return (int) floor((16 ** $expectedCodeLength) * $this->ratio - $generatedAmount);
+        $codeCombination = 16 ** $expectedCodeLength * $this->ratio;
+        if ($codeCombination >= PHP_INT_MAX) {
+            return PHP_INT_MAX - $generatedAmount;
+        }
+
+        return (int) $codeCombination - $generatedAmount;
```

Comment:
Example:
When: `$expectedAmount = 20;`
Then: `$codeCombination = {float} 6.0446290980731E+23;`
And: `$codeCombination >= PHP_INT_MAX` is true
So the value `PHP_INT_MAX - $generatedAmount` will be returned

IMO this fix (without changing current behaviour and without BC breaks) makes sense, but WDYT /cc @Sylius/core-team ? 😃

Reply:
In PromotionCouponGenerator we are allowing for coupons up to 40 characters. If we can process with calculation, maybe we should increase this amount? 

Also, how much of precision will we lost when we will convert to float?

Reply:
The problem is that we cannot convert to float because of BC Policy, we need to return here an integer value, so `PHP_INT_MAX` is the max value that could be returned. IMO this fix resolves the bug, we are able to generate coupons even for 40 characters, but there will be a limit to `PHP_INT_MAX` coupons that can be generated. Increasing this value could be also a problem for form (`PromotionCouponGeneratorInstructionType`) where is an integer field for the amount.

Reply:
Thanks for the explanation!

Reply:
Thanks for the explanation!